File: TakeFive/takeFive/src/main/java/com/teuskim/takefive/SettingsActivity.java
package com.teuskim.takefive;

import android.os.Bundle;
import android.text.Html;
import android.view.View;
import android.widget.TextView;

import com.teuskim.takefive.common.OmokPreference;
import com.teuskim.takefive.iab.Purchase;

public class SettingsActivity extends BaseGameActivity {
	
	private ViewControllaer viewCtrl;
	private OmokPreference pref;
	
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.settings_activity);
		viewCtrl = new ViewControllaer();
		viewCtrl.init();
		pref = OmokPreference.getInstance(getApplicationContext());
		initStates();
	}
	
	private void initStates() {
		viewCtrl.setBgm(pref.isOnBgm());
		viewCtrl.setGameSound(pref.isOnGameSound());
		viewCtrl.setInvitedOk(pref.isOnInvitedOk());
		viewCtrl.setFirstStageHelper(pref.isOnFirstStageHelper());
	}
	
//	@Override
//	protected void afterIab(Purchase item) {
//		super.afterIab(item);
//
//		if (item == null) {
//			viewCtrl.showItemView();
//		} else {
//			viewCtrl.hideItemView();
//		}
//	}

	private void toggleBgm() {
		boolean isOn = !viewCtrl.bgmOn.isShown();
		viewCtrl.setBgm(isOn);
		pref.setIsOnBgm(isOn);
		if (isOn) {
			startBgmForMenu();
			googleAnalyticsEvent(GA_ACTION_BUTTON_CLICK, GA_LABEL_BGM_ON);
		} else {
			pauseBgm();
			googleAnalyticsEvent(GA_ACTION_BUTTON_CLICK, GA_LABEL_BGM_OFF);
		}
	}

	private void toggleGameSound() {
		boolean isOn = !viewCtrl.gameSoundOn.isShown();
		viewCtrl.setGameSound(isOn);
		pref.setIsOnGameSound(isOn);
		googleAnalyticsEvent(GA_ACTION_BUTTON_CLICK, isOn ? GA_LABEL_GAME_SOUND_ON : GA_LABEL_GAME_SOUND_OFF);
	}

	private void toggleInvitedOk() {
		boolean isOn = !viewCtrl.invitedOkOn.isShown();
		viewCtrl.setInvitedOk(isOn);
		pref.setIsOnInvitedOk(isOn);
		googleAnalyticsEvent(GA_ACTION_BUTTON_CLICK, isOn ? GA_LABEL_INVITED_OK_ON : GA_LABEL_INVITED_OK_OFF);
	}
	
	private void toggleFirstStageHelper() {
		boolean isOn = !viewCtrl.firstStageHelperOn.isShown();
		viewCtrl.setFirstStageHelper(isOn);
		pref.setIsOnFirstStageHelper(isOn);
		googleAnalyticsEvent(GA_ACTION_BUTTON_CLICK, isOn ? GA_LABEL_FIRST_STAGE_HELPER_ON : GA_LABEL_FIRST_STAGE_HELPER_OFF);
	}
	

	class ViewControllaer {
		private View bgmOn, bgmOff;
		private View gameSoundOn, gameSoundOff;
		private View invitedOkOn, invitedOkOff;
		private View firstStageHelperOn, firstStageHelperOff;
		private TextView textBuyItemView;
		private View btnBuyItem;
		
		public void init() {
			View.OnClickListener listener = new View.OnClickListener() {
				
				@Override
				public void onClick(View v) {
					playButtonClick();
					
					switch (v.getId()) {
					case R.id.btn_bgm:
						toggleBgm();
						break;
					case R.id.btn_game_sound:
						toggleGameSound();
						break;
					case R.id.btn_invited_ok:
						toggleInvitedOk();
						break;
					case R.id.btn_first_stage_helper:
						toggleFirstStageHelper();
						break;
					case R.id.btn_buy_item:
//						openBuyItem();
						googleAnalyticsEvent(GA_ACTION_BUTTON_CLICK, GA_LABEL_BUY_ITEM);
						break;
					}
				}
			};
			findViewById(R.id.btn_bgm).setOnClickListener(listener);
			bgmOn = findViewById(R.id.bgm_on);
			bgmOff = findViewById(R.id.bgm_off);
			findViewById(R.id.btn_game_sound).setOnClickListener(listener);
			gameSoundOn = findViewById(R.id.game_sound_on);
			gameSoundOff = findViewById(R.id.game_sound_off);
			findViewById(R.id.btn_invited_ok).setOnClickListener(listener);
			invitedOkOn = findViewById(R.id.invited_ok_on);
			invitedOkOff = findViewById(R.id.invited_ok_off);
			findViewById(R.id.btn_first_stage_helper).setOnClickListener(listener);
			firstStageHelperOn = findViewById(R.id.first_stage_helper_on);
			firstStageHelperOff = findViewById(R.id.first_stage_helper_off);
			
			textBuyItemView = (TextView) findViewById(R.id.text_buy_item);
			textBuyItemView.setText(Html.fromHtml(getString(R.string.main_message_6)));
			btnBuyItem = findViewById(R.id.btn_buy_item);
			btnBuyItem.setOnClickListener(listener);
			hideItemView();
		}
		
		public void setBgm(boolean isOn) {
			if (isOn) {
				bgmOn.setVisibility(View.VISIBLE);
				bgmOff.setVisibility(View.GONE);
			} else {
				bgmOn.setVisibility(View.GONE);
				bgmOff.setVisibility(View.VISIBLE);
			}
		}
		
		public void setGameSound(boolean isOn) {
			if (isOn) {
				gameSoundOn.setVisibility(View.VISIBLE);
				gameSoundOff.setVisibility(View.GONE);
			} else {
				gameSoundOn.setVisibility(View.GONE);
				gameSoundOff.setVisibility(View.VISIBLE);
			}
		}
		
		public void setInvitedOk(boolean isOn) {
			if (isOn) {
				invitedOkOn.setVisibility(View.VISIBLE);
				invitedOkOff.setVisibility(View.GONE);
			} else {
				invitedOkOn.setVisibility(View.GONE);
				invitedOkOff.setVisibility(View.VISIBLE);
			}
		}
		
		public void setFirstStageHelper(boolean isOn) {
			if (isOn) {
				firstStageHelperOn.setVisibility(View.VISIBLE);
				firstStageHelperOff.setVisibility(View.GONE);
			} else {
				firstStageHelperOn.setVisibility(View.GONE);
				firstStageHelperOff.setVisibility(View.VISIBLE);
			}
		}
		
		public void hideItemView() {
			textBuyItemView.setVisibility(View.GONE);
			btnBuyItem.setVisibility(View.GONE);
		}
		
		public void showItemView() {
			textBuyItemView.setVisibility(View.VISIBLE);
			btnBuyItem.setVisibility(View.VISIBLE);
		}
	}

}

File: TakeFive_kotlin/takeFive/src/main/java/com/teuskim/takefive/SettingsActivity.kt
package com.teuskim.takefive

import android.os.Bundle
import android.text.Html
import android.view.View
import android.widget.TextView

import com.teuskim.takefive.common.OmokPreference
//import com.teuskim.takefive.iab.Purchase
import com.teuskim.takefive.common.IGoogleAnalytics

class SettingsActivity : BaseGameActivity() {

    private var viewCtrl: ViewControllaer? = null
    private var pref: OmokPreference? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.settings_activity)
        viewCtrl = ViewControllaer()
        viewCtrl!!.init()
        pref = OmokPreference.getInstance(applicationContext)
        initStates()
    }

    private fun initStates() {
        viewCtrl!!.setBgm(pref!!.isOnBgm)
        viewCtrl!!.setGameSound(pref!!.isOnGameSound)
        viewCtrl!!.setInvitedOk(pref!!.isOnInvitedOk)
        viewCtrl!!.setFirstStageHelper(pref!!.isOnFirstStageHelper)
    }

//    override fun afterIab(item: Purchase?) {
//        super.afterIab(item)
//
//        if (item == null) {
//            viewCtrl!!.showItemView()
//        } else {
//            viewCtrl!!.hideItemView()
//        }
//    }

    private fun toggleBgm() {
        val isOn = !viewCtrl!!.bgmOn!!.isShown
        viewCtrl!!.setBgm(isOn)
        pref!!.isOnBgm = isOn
        if (isOn) {
            startBgmForMenu()
            googleAnalyticsEvent(IGoogleAnalytics.GA_ACTION_BUTTON_CLICK, IGoogleAnalytics.GA_LABEL_BGM_ON)
        } else {
            pauseBgm()
            googleAnalyticsEvent(IGoogleAnalytics.GA_ACTION_BUTTON_CLICK, IGoogleAnalytics.GA_LABEL_BGM_OFF)
        }
    }

    private fun toggleGameSound() {
        val isOn = !viewCtrl!!.gameSoundOn!!.isShown
        viewCtrl!!.setGameSound(isOn)
        pref!!.isOnGameSound = isOn
        googleAnalyticsEvent(IGoogleAnalytics.GA_ACTION_BUTTON_CLICK, if (isOn) IGoogleAnalytics.GA_LABEL_GAME_SOUND_ON else IGoogleAnalytics.GA_LABEL_GAME_SOUND_OFF)
    }

    private fun toggleInvitedOk() {
        val isOn = !viewCtrl!!.invitedOkOn!!.isShown
        viewCtrl!!.setInvitedOk(isOn)
        pref!!.isOnInvitedOk = isOn
        googleAnalyticsEvent(IGoogleAnalytics.GA_ACTION_BUTTON_CLICK, if (isOn) IGoogleAnalytics.GA_LABEL_INVITED_OK_ON else IGoogleAnalytics.GA_LABEL_INVITED_OK_OFF)
    }

    private fun toggleFirstStageHelper() {
        val isOn = !viewCtrl!!.firstStageHelperOn!!.isShown
        viewCtrl!!.setFirstStageHelper(isOn)
        pref!!.isOnFirstStageHelper = isOn
        googleAnalyticsEvent(IGoogleAnalytics.GA_ACTION_BUTTON_CLICK, if (isOn) IGoogleAnalytics.GA_LABEL_FIRST_STAGE_HELPER_ON else IGoogleAnalytics.GA_LABEL_FIRST_STAGE_HELPER_OFF)
    }


    internal inner class ViewControllaer {
        var bgmOn: View? = null
        private var bgmOff: View? = null
        var gameSoundOn: View? = null
        private var gameSoundOff: View? = null
        var invitedOkOn: View? = null
        private var invitedOkOff: View? = null
        var firstStageHelperOn: View? = null
        private var firstStageHelperOff: View? = null
        private var textBuyItemView: TextView? = null
        private var btnBuyItem: View? = null

        fun init() {
            val listener = View.OnClickListener { v ->
                playButtonClick()

                when (v.id) {
                    R.id.btn_bgm -> toggleBgm()
                    R.id.btn_game_sound -> toggleGameSound()
                    R.id.btn_invited_ok -> toggleInvitedOk()
                    R.id.btn_first_stage_helper -> toggleFirstStageHelper()
                    R.id.btn_buy_item ->
                        //						openBuyItem();
                        googleAnalyticsEvent(IGoogleAnalytics.GA_ACTION_BUTTON_CLICK, IGoogleAnalytics.GA_LABEL_BUY_ITEM)
                }
            }
            findViewById<View>(R.id.btn_bgm).setOnClickListener(listener)
            bgmOn = findViewById(R.id.bgm_on)
            bgmOff = findViewById(R.id.bgm_off)
            findViewById<View>(R.id.btn_game_sound).setOnClickListener(listener)
            gameSoundOn = findViewById(R.id.game_sound_on)
            gameSoundOff = findViewById(R.id.game_sound_off)
            findViewById<View>(R.id.btn_invited_ok).setOnClickListener(listener)
            invitedOkOn = findViewById(R.id.invited_ok_on)
            invitedOkOff = findViewById(R.id.invited_ok_off)
            findViewById<View>(R.id.btn_first_stage_helper).setOnClickListener(listener)
            firstStageHelperOn = findViewById(R.id.first_stage_helper_on)
            firstStageHelperOff = findViewById(R.id.first_stage_helper_off)

            textBuyItemView = findViewById<View>(R.id.text_buy_item) as TextView
            textBuyItemView!!.text = Html.fromHtml(getString(R.string.main_message_6))
            btnBuyItem = findViewById(R.id.btn_buy_item)
            btnBuyItem!!.setOnClickListener(listener)
            hideItemView()
        }

        fun setBgm(isOn: Boolean) {
            if (isOn) {
                bgmOn!!.visibility = View.VISIBLE
                bgmOff!!.visibility = View.GONE
            } else {
                bgmOn!!.visibility = View.GONE
                bgmOff!!.visibility = View.VISIBLE
            }
        }

        fun setGameSound(isOn: Boolean) {
            if (isOn) {
                gameSoundOn!!.visibility = View.VISIBLE
                gameSoundOff!!.visibility = View.GONE
            } else {
                gameSoundOn!!.visibility = View.GONE
                gameSoundOff!!.visibility = View.VISIBLE
            }
        }

        fun setInvitedOk(isOn: Boolean) {
            if (isOn) {
                invitedOkOn!!.visibility = View.VISIBLE
                invitedOkOff!!.visibility = View.GONE
            } else {
                invitedOkOn!!.visibility = View.GONE
                invitedOkOff!!.visibility = View.VISIBLE
            }
        }

        fun setFirstStageHelper(isOn: Boolean) {
            if (isOn) {
                firstStageHelperOn!!.visibility = View.VISIBLE
                firstStageHelperOff!!.visibility = View.GONE
            } else {
                firstStageHelperOn!!.visibility = View.GONE
                firstStageHelperOff!!.visibility = View.VISIBLE
            }
        }

        fun hideItemView() {
            textBuyItemView!!.visibility = View.GONE
            btnBuyItem!!.visibility = View.GONE
        }

        fun showItemView() {
            textBuyItemView!!.visibility = View.VISIBLE
            btnBuyItem!!.visibility = View.VISIBLE
        }
    }

}
